src/workerd/jsg/modules-new.c++

```diff
@@ -0,0 +1,1401 @@
+          return;
+        }
+
+        // The import.meta.resolve(...) function is effectively a shortcut for
+        // new URL(specifier, import.meta.url).href. The idea is that it allows
+        // resolving import specifiers relative to the current modules base URL.
+        // Note that we do not validate that the resolved URL actually matches
+        // anything in the registry.
+        auto resolve = js.wrapReturningFunction(js.v8Context(),
+            [href=kj::mv(href)]
+            (Lock& js, const v8::FunctionCallbackInfo<v8::Value>& args) -> JsValue {
+          // Note that we intentionally use ToString here to coerce whatever value is given
+          // into a string or throw if it cannot be coerced.
+          auto specifier = js.toString(args[0]);
+          KJ_IF_SOME(resolved, Url::tryParse(specifier.asPtr(), href)) {
+            auto normalized = resolved.clone(Url::EquivalenceOption::NORMALIZE_PATH);
+            return js.str(normalized.getHref());
+          } else {
+            // If the specifier could not be parsed and resolved successfully,
+            // the spec says to return null.
+            return js.null();
+          }
+        });
+
+        if (meta->CreateDataProperty(js.v8Context(),
+            v8::Local<v8::String>(js.strIntern("resolve"_kj)),
+            resolve).IsNothing()) {
+          return;
+        }
+      }
+    }, [&](Value exception) {
+      // It would be exceedingly odd to end up here but we handle it anyway,
+      // just to ensure that we do not crash the isolate. The only thing we'll
+      // do is rethrow the error.
+      js.v8Isolate->ThrowException(exception.getHandle(js));
+    });
+  } catch(...) {
+    kj::throwFatalException(kj::getCaughtExceptionAsKj());
+  }
+}
+
+// The callback v8 calls when dynamic import(...) is used.
+v8::MaybeLocal<v8::Promise> dynamicImport(
+    v8::Local<v8::Context> context,
+    v8::Local<v8::Data> host_defined_options,
+    v8::Local<v8::Value> resource_name,
+    v8::Local<v8::String> specifier,
+    v8::Local<v8::FixedArray> import_assertions) {
+  auto isolate = context->GetIsolate();
+  auto& js = Lock::from(isolate);
+  try {
+    return js.tryCatch([&]() -> v8::MaybeLocal<v8::Promise> {
+      auto spec = js.toString(specifier);
+
+      // The proposed specification for import assertions strongly recommends that
+      // embedders reject import attributes and types they do not understand/implement.
+      // This is because import attributes can alter the interpretation of a module.
+      // Throwing an error for things we do not understand is the safest thing to do
+      // for backwards compatibility.
+      //
+      // For now, we do not support any import attributes, so if there are any at all
+      // we will reject the import.
+      if (!import_assertions.IsEmpty()) {
+        if (import_assertions->Length() > 0) {
+          auto error = js.typeError("Import attributes are not supported");
```

Comment:
Nit: This pattern to return a rejected promise with an error message appears 3 times in this function – would it make sense to define a lambda for it?